### DISCUSSION TITLE
[firebase_storage] Is it possible to delete a folder?

### DISCUSSION THREAD

**Discussion by tgbarker** [firebase/flutterfire#6066] [Q&A]
firebase_storage: ^8.0.3

I have a function that I thought would work, but does not.

```
  @override
  Future<void> deleteFolder(String filePath) async {
    return await firebase_storage
        .FirebaseStorage.instance
        .ref('$filePath')
        .delete();
}
```
The filePath is correct and points to a folder in my storage which contains a few sub folders and files.

E/StorageException(19263): StorageException has occurred.
E/StorageException(19263): Object does not exist at location.
E/StorageException(19263):  Code: -13010 HttpResult: 404
E/StorageException(19263): {  "error": {    "code": 404,    "message": "Not Found.  Could not delete object",    "status": "DELETE_OBJECT"  }}
E/StorageException(19263): java.io.IOException: {  "error": {    "code": 404,    "message": "Not Found.  Could not delete object",    "status": "DELETE_OBJECT"  }}

Imagine I have the file structure

parent/child1/subfolder1

I would like to delete the child1 folder. Is this possible? I saw a listAll() function, surely I don't have to loop by hand with nested ListResult returns?

**Comment by tgbarker** [comment]
So anyone else looking for a solution, like @MiguelSOliveira and the duplicate question posted after mine. 

Directories do not really exist and the "folder structure" is not really a classic true folder. If you delete all the files, then the folder no longer exists in the firebase admin console. So anyone can think about the following code which is a recursive function to delete all files from a "folder" (essentially a root filePath).

```
 Future<void> deleteFolder(String filePath) async {
    logger.d("Deleting storage folder for : " + filePath);
    firebase_storage.ListResult res =
        await firebase_storage.FirebaseStorage.instance
            .ref('$filePath')
            //.delete();
            .listAll();
    //await firebase_storage.FirebaseStorage.instance.ref('$filePath').
    for (firebase_storage.Reference item in res.items) {
      logger.d("Deleting : " + item.name);
      await firebase_storage.FirebaseStorage.instance
          .ref(item.fullPath)
          .delete();
    }
    for (firebase_storage.Reference item in res.prefixes) {
      logger.d("prefix : " + item.name);
      deleteFolder(item.fullPath);
    }
  }
```

> I saw a listAll() function, surely I don't have to loop by hand with nested ListResult returns?
Yes, it seems I do have to.

**Reply by MiguelSOliveira** [reply]
Thanks for the reply @tgbarker .

`If you delete all the files, then the folder no longer exists in the firebase admin console.`

The folder is still present in firebase storage though, so it wasn't really deleted. That is my current problem.

**Reply by tgbarker** [reply]
> 
> 
> Thanks for the reply @tgbarker .
> 
> `If you delete all the files, then the folder no longer exists in the firebase admin console.`
> 
> The folder is still present in firebase storage though, so it wasn't really deleted. That is my current problem.

That is strange because when I did the testing, if I deleted all the files under a "folder" and refreshed the admin console then the "folder" no longer exists.

**Reply by MiguelSOliveira** [reply]
Maybe it's because i'm trying to delete the root folder? In your case would be deleting the "parentFolder".

I want to delete all the way to top level folders.
